module/zfs/dbuf.c

```diff
@@ -1167,50 +1222,327 @@ dbuf_whichblock(const dnode_t *dn, const int64_t level, const uint64_t offset)
 	}
 }
 
+typedef struct dbuf_dirty_record_hole {
+	caddr_t src;
+	caddr_t dst;
+	int size;
+} dbuf_dirty_record_hole_t;
+
+typedef struct dbuf_dirty_record_hole_itr {
+	/* provided data */
+	arc_buf_t *src;
+	dbuf_dirty_leaf_record_t *dl;
+	/* calculated data */
+	dbuf_dirty_range_t *range;
+	/* One greater than the last valid offset in the dst buffer */
+	int max_offset;
+	int hole_start;
+	dbuf_dirty_record_hole_t hole;
+} dbuf_dirty_record_hole_itr_t;
```

Comment:
These structures, and their associated machinery, are not useful separate from COW avoidance.  Their purpose is to track which parts of the dirty record are untouched (and therefore are waiting for a resolving asynchronous read of the block's last txg), which is a meaningless concept if writes can only modify a fully loaded dirty record.

I'm not sure it's realistic or worthwhile to try to break out a refactoring of `dbuf_read_done` in this commit from #9954.